stripped/colabs/Implementacja_NearestNeighbours_stripped.py

```diff
@@ -0,0 +1,177 @@
+"""
+
+!apt-get install openjdk-8-jdk-headless -qq > /dev/null
+!wget -q dlcdn.apache.org/spark/spark-3.5.0/spark-3.5.0-bin-hadoop3.tgz
+!tar xf spark-3.5.0-bin-hadoop3.tgz
+!pip install -q findspark
+
+
+import os
+os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
+os.environ["SPARK_HOME"] = "/content/spark-3.5.0-bin-hadoop3"
+
+import findspark
+findspark.init()
+
+
+from functools import reduce
+import math
+from typing import Optional, List, Tuple
+
+import pandas as pd
+import numpy as np
+from geopy.distance import geodesic
+
+from pyspark.sql import SparkSession, DataFrame as SparkDataFrame, functions as F
+from pyspark.sql.types import IntegerType, FloatType, StringType, StructType, StructField, Row
+
+from google.colab import drive
+
+
+spark = (
+         SparkSession.builder
+        .master("local")
+        .appName("Colab")
+        .config('spark.ui.port', '4050')
+        .getOrCreate()
+)
+conf = spark.sparkContext._conf.setAll([('spark.executor.memory', '100g'), ('spark.driver.memory','64g')])
+spark.conf.set("park.driver.maxResultSize", "80g")
+
+spark.conf.set('spark.sql.execution.arrow.enabled', 'true')
+
+
+drive.mount('/content/drive')
+
+
+# Wczytanie zbioru sampled w celu pobrania nazw kolumn
+sampled = pd.read_csv('/content/drive/MyDrive/BigMess/NASA/sampled_NASA_200k.csv')
+
+# Utworzenie schematu określającego typ zmiennych
+schemat = StructType()
+for i in sampled.columns:
+  if i == "Date":
+    schemat = schemat.add(i, IntegerType(), True)
+  else:
+    schemat = schemat.add(i, FloatType(), True)
+
+
+%%time
+# Wczytanie zbioru Nasa w sparku
+nasa = spark.read.format('csv').option("header", True).schema(schemat).load('/content/drive/MyDrive/BigMess/NASA/NASA.csv')
+nasa.show(5)
+
+nasa.createOrReplaceTempView("nasa")
+
+nasa = (
+    nasa
+    .withColumn('Year', (F.col('Date') / 100).cast('int'))
+    .withColumn('Month', F.col('Date') % 100)
+    .drop('Date')
+)
+nasa.show(5)
+
+
+nasa_coords = spark.sql("""SELECT DISTINCT lat, lon FROM nasa""")
+nasa_coords.collect()
+
+
+"""
+### **2. Implementacja algorytmu kNN:**
+"""
+
+# function searches for points that lie within a (euclidean) ball of size *radius* around the query points
+# if optional argument *k* is given then function searches for (at most) k nearest points that lie within
+# a (euclidean) ball of size *radius* around the query points
+
+#point = (latitude, longitude)
+def kRadiusNN(df: SparkDataFrame, radius: float, point: Tuple[float, float], k: Optional[int]=None) -> pd.DataFrame :
+  assert (25.0625<= point[0] <=52.9375) and (-124.9375<= point[1] <=-67.0625 ), 'Wrong coordinates (out of range)'
+
+  df = df.collect()
+  neighbours_pd = pd.DataFrame({"lon":[], "lat":[], 'dist':[]})
+
+  for row in df:
+
+     lon = row['lon']
+     lat = row['lat']
+     dist = geodesic((lat, lon), point).km
+     if dist <= radius:
+        new_row = {'lon': lon , 'lat': lat, 'dist' : dist}
+        neighbours_pd.loc[len(neighbours_pd)] = new_row
```

Comment:
to jest słabe wydajnościowo. Dużo lepiej z perspektywy Big Data byłoby utworzyć sobie w df nową kolumnę gdzie mamy odległość od point. Dalej trochę jest wymyślania koła na nowo. Do zrobienia klasyfikatora większościowego wystarczyłoby 
1. utworzyć kolumnę (można przekazać lambdę albo np. użyć funkcji udf https://stackoverflow.com/questions/70315069/calculate-the-geographical-distance-in-pyspark-dataframe)
2. odfiltrować tam gdzie nowa kolumna i radius spełniają odległość
3. Użyć pyspark.sql.functions as F - F.mode
4. Problemem może być typowanie, ale do takich sytuacji w razie czego mamy możliwość typować Union[str,List[str]] lub coś podobnego

Reply:
Problem zaadresowany w korespondencji na MS Teams.